shared/src/main/scala/mlscript/ucs/Desugarer.scala

```diff
@@ -60,7 +60,7 @@ class Desugarer extends TypeDefs { self: Typer =>
     }.toList
   }
 
-  private val localizedScrutineeMap = MutMap.empty[Term, Var]
+  private val localizedScrutineeMap = MutMap.empty[(Opt[Loc], Term), Var]
```

Comment:
I still hate this. What if a scrutinee doesn't have a location for some reason (eg. maybe it's synthetic)? The robustness of your code SHOULD NOT depend on locations being present in the trees!!

Reply:
Sure. Let me think what I can do.

Reply:
If you really can't think of anything else, you can always try a `java.util.IdentityHashMap` https://docs.oracle.com/javase/7/docs/api/java/util/IdentityHashMap.html

Reply:
Actually `IdentityHashMap` doesn’t work in my case.

Reply:
The goal is to localize non-`Var` scrutinees by creating let bindings so that we can reduce unnecessary computations. For example, we'd like to translate
```
if foo(x, y, z)
  < 0 then left
  > 0 then right
  = _ then this
```
to
```
let t = foo(x, y, z) in
case t < 0 of
  true -> left
  false -> case t > 0 of
    true -> right
    false -> this
```
rather than
```
case foo(x, y, z) < 0 of
  true -> left
  false -> case foo(x, y, z) > 0 of
    true -> right
    false -> this
```

Reply:
Why wouldn't `IdentityHashMap` work? I believe it does precisely what you want.

Reply:
Actually I'm not really clear on why you need a map at all, but if you really do `IdentityHashMap` should work as we're talking about a specific non-duplicated term here. But my feeling is that you're doing something in a suboptimal way.